src/concrete/ml/quantization/quantized_ops.py

```diff
@@ -1577,25 +1577,288 @@ class QuantizedOr(QuantizedOpUnivariateOfEncrypted, QuantizedOp):
+        assert self.output_quant_params.zero_point is not None
+
+        prepared_inputs = self._prepare_inputs_with_constants(
+            *q_inputs, calibrate=False, quantize_actual_values=True
+        )
+
+        q_input_0: QuantizedArray = prepared_inputs[0]
+        q_input_1: QuantizedArray = prepared_inputs[1]
+
+        assert q_input_0.quantizer.scale is not None
+        assert q_input_0.quantizer.zero_point is not None
+
+        assert q_input_1.quantizer.scale is not None
+        assert q_input_1.quantizer.zero_point is not None
+
+        # Remove the manual encrypted multiplication when we
+        # can handle input precision with rounding
+        # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4127
+        @univariate
+        def copy_function(x):
+            return x
+
+        input0_q_values = q_input_0.qvalues
+        input1_q_values = q_input_1.qvalues
+
+        with tag(self.op_instance_name + ".enc_mul_rounding_input0"):
+            input0_q_values = self.cnp_round(
+                input0_q_values, calibrate_rounding, rounding_operation_id="input0_q_values_copy"
+            )
+
+        with tag(self.op_instance_name + ".enc_mul_rounding_input1"):
+            input1_q_values = self.cnp_round(
+                input1_q_values, calibrate_rounding, rounding_operation_id="input1_q_values_copy"
+            )
+
+        input0_q_values = copy_function(input0_q_values)
+        input1_q_values = copy_function(input1_q_values)
+
+        # The product of quantized encrypted integer values
+        product_q_values = input0_q_values * input1_q_values
+
+        # Integer quantized multiplication need adjustment based on the zero points.
+        if q_input_0.quantizer.zero_point:
+            product_q_values -= q_input_0.quantizer.zero_point * input1_q_values
+        if q_input_1.quantizer.zero_point:
+            product_q_values -= q_input_1.quantizer.zero_point * input0_q_values
+
+        # Compute the scale and zero point based on the scale and zero point
+        # of the two quantized values multiplied together
+        new_scale = q_input_0.quantizer.scale * q_input_1.quantizer.scale
+        new_zero_point = q_input_0.quantizer.zero_point * q_input_1.quantizer.zero_point
+
+        if self.produces_graph_output:
+            return self.make_output_quant_parameters(product_q_values, new_scale, new_zero_point)
+
+        with tag(self.op_instance_name + ".rounding"):
+            # Apply Concrete rounding (if relevant)
+            product_q_values = self.cnp_round(
+                product_q_values, calibrate_rounding, rounding_operation_id="product_q_values"
+            )
+
+        # De-quantize the product
+        dequant_product = (product_q_values + new_zero_point) * new_scale
+
+        # Return the raw float values without re-quantizing them to the new scale, as any
+        # following Gemm/Add/Conv will quantize them with _prepare_inputs_with_constants(...)
```

Comment:
same comment as above, no issue if at the end of the graph or before any reshape/squeeze ect operators ?

Reply:
same answer we have a if condition to return the integer if needed

Reply:
no, this is standard